### PR TITLE
fix: account for DSA top-k sparsity and indexer cost in FLOPs

### PR DESCRIPTION
**Pull request by niyunsheng**
Port of #6753 (merged to `dev`) onto `main`, extended with **hybrid-path DSA accounting** — which on `main` is the part that actually matters.

## Problem

On `main`, DSA models (GLM-5.2) are only buildable through the hybrid path (`--hybrid-layer-pattern` with `D` layers; the standard GPT path rejects experimental attention variants). `hybrid_flops` counts only `M`/`G`/`*`/`-`/`E` layers, so every `D` layer — the absorbed-MLA attention *and* its indexer — is silently missing from `num_floating_point_operations`. For a small DSA proxy (16 hybrid layers, seq 4096) the reported throughput is 2.26x too low; the gap grows with the attention share of the model.

The dev PR instead guarded the hybrid path with a fail-loud assert. That is not portable to `main`: with hybrid as the only DSA path and `num_floating_point_operations` running unconditionally every iteration, the assert would abort every DSA training run.

## Changes

`megatron/training/training.py`:

- **Hybrid path (new here, not in the dev PR):** count `D` layers on top of `hybrid_flops` — per-layer absorbed-MLA projections and top-k-scaled sparse core attention (`_dsa_layer_flops`), plus the lightning indexer on the `D` positions that compute their own top-k index (`_num_dsa_indexer_layers_in_pattern`). The indexer count uses the block-local layer numbers `DSAttention` actually receives: the main pattern numbers layers sequentially across `|` pipes, and each hybrid MTP depth restarts at 1 (`pp_layer_offset=0`). The indexer keeps its own KL-loss-dependent fwd/bwd expansion (1x off; 2x projections when on; scoring 3x with the dense KL or `1 + 2 * sparse_scale` with the sparse KL, see below), so it is added after `hybrid_flops`'s global x3.
- **Standard path (ported verbatim from #6753):** the `"dsa"` branch in `transformer_flops` — absorbed-MLA core form, top-k sparsity scale, indexer. Unreachable on `main` today (the experimental-variant spec helper has no callers yet), but it keeps this PR line-for-line with the dev-reviewed change and becomes live as soon as standard-path DSA support lands. See #6753 for the derivation and full-scale GLM-5.2 measurements.

`tests/unit_tests/test_num_floating_point_operations.py`: the dev PR's `TestDSA`/`TestDSAHelperEdgeCases` (BSHD/THD goldens, loss on/off expansions, cross-layer sharing, top-k caps long-context growth), plus `TestHybridDSA`: since `D` layers contributed nothing before, deleting them from the pattern isolates exactly the DSA terms — the delta is pinned against an independent golden; cross-layer sharing is pinned on block-local numbering, and an MTP case pins that depths restart at 1.

## Test plan

- Unit tests on a lyris GB200 node under the CI harness (`Dockerfile.ci.dev`, `torch.distributed.run --nproc_per_node 4`): results in a comment below.
- **Numerical equivalence**: one fixed 16-layer DSA arg set (`ffn_hidden_size=0`, no MoE/MTP, so the comparable surface is exactly this PR's terms) fed to `dev`@85902ef standard path, this PR's standard path, and this PR's hybrid path (`"D"*16`) returns **bit-identical** FLOPs (`146585086328832.0`) — the port is faithful and the hybrid accounting reproduces the dev-reviewed formulas exactly.
- **End-to-end smoke at this head**: a hybrid DSA proxy (`--hybrid-layer-pattern D-D-DEDEDEDEDEDE`, DSv3-proxy MLA dims, MoE 32E, indexer loss on) trained 10 iterations on 4 GB200 GPUs via `pretrain_hybrid.py`; stable TFLOP/s/GPU from the new accounting (20.5 vs 9.1 before this PR).
- `tools/autoformat.sh` (black 26.3.0) and `tools/check_copyright.py` pass on the changed files.

## Update: honour `--dsa-indexer-use-sparse-loss` (second commit)

Counterpart of #7157 on `dev`; the change is identical except that here both the standard-path and the hybrid-path call sites are updated.

#6753 charged the indexer scoring pass a flat `fwd + dq + dk = 3x` whenever `dsa_indexer_loss_coeff > 0`, and its docstring noted that the sparse KL variant was not modelled. With `--dsa-indexer-use-sparse-loss` the KL is taken over the selected top-k keys only, so the score gradient is zero outside them and dq/dk of the scoring GEMM span `sum_i(min(i, topk))` pairs instead of the dense causal `L^2 / 2`. The fused cuDNN backend (`_compute_sparse_indexer_loss_and_grads` -> `_run_sparse_indexer_backward`) executes exactly this sparse backward; the reference PyTorch path multiplies a dense zero-padded gradient, which is an implementation detail and, like dense-masked reference attention, not part of the model FLOPs count. Since the indexer's dense O(L^2) scoring dominates the quadratic term at long context (core attention is already top-k scaled), the overcount on that term was close to 3x.

- `_dsa_indexer_flops` takes `dsa_indexer_use_sparse_loss` and `sparse_core_scale` (the same top-k / dense pair ratio `_dsa_sparse_core_scale` applies to core attention). Scoring expansion: indexer loss off 1x (unchanged); dense KL 3x (unchanged); sparse KL `1 + 2 * sparse_core_scale`.
- Parameters renamed to match the corresponding `args` (`dsa_indexer_loss_coeff`, `dsa_indexer_use_sparse_loss`). Both the standard-path and the hybrid-path (`D` layers) call sites pass the flag and the shared scale.
- Default configs (`dsa_indexer_use_sparse_loss=False`) produce identical numbers to before.
- Tests: golden calculator follows the same rule; `TestDSA::test_sparse_indexer_loss_shrinks_scoring_backward[256|8192]` (sparse == dense at `L <= topk`, strictly smaller past topk and still above a frozen indexer), `TestHybridDSA::test_sparse_indexer_loss_shrinks_scoring_backward` (same on the hybrid call site), `test_sparse_loss_ignored_without_indexer_loss`, and `test_indexer_flops_sparse_loss_expansion` (direct 1x / 3x / `1 + 2 * scale` check).
- Unit tests rerun on a lyris GB200 node under the CI harness: `42 passed, 8 skipped` (skips are the 8-rank topology cases on a 4-GPU node). `black` 26.3.0 clean.

Not covered, consistent with the rule of excluding auxiliary-loss work: the teacher-side attention score recompute inside the indexer KL loss (dense over all keys, or top-k only under the sparse variant).

### CODE REVIEW

**Review by zhongbozhu — APPROVED** [review]

File: megatron/training/training.py [hunk 1]
```diff
@@ -818,6 +821,227 @@ def consume_seqlen_stats_in_iteration() -> Tuple[Optional[float], Optional[float
     return total_real_tokens / dedup, seqlen_squared_sum / dedup
 
 
+def _dsa_sparse_core_scale(total_real_tokens, seqlen_squared_sum, dsa_indexer_topk):
```
Comment (deepakn94) [inline megatron/training/training.py:824]: Can we move these new FLOPs functions and the existing ones into a separate new file in `megatron/training`? These functions are getting too long and `megatron/training/training.py` needs to be cleaned.

Comment (niyunsheng) [reply megatron/training/training.py:824]: @deepakn94 Agreed that the FLOPs calculations should live in a separate module. Since this PR fixes DSA FLOPs accounting, would you prefer a separate refactoring PR to move all FLOPs functions into `megatron/training/flops.py`? That would keep the correctness changes easier to review. I’m happy to do the extraction as a follow-up, or land it first and rebase this PR if you prefer.

File: megatron/training/training.py [hunk 2]
```diff
@@ -818,6 +821,228 @@ def consume_seqlen_stats_in_iteration() -> Tuple[Optional[float], Optional[float
...
+):
+    """DSA lightning-indexer FLOPs coefficients, fwd/bwd expansion included.
+
+    Counts the indexer's projections (a Q projection off the shared ``q_lora``
+    residual, the ``linear_wk`` key path, and the per-head ``weights_proj``)
+    plus its dense scoring pass of every query against every past token under
+    a causal mask. Scoring is ``O(L^2)`` even though the attention consuming
+    it is sparse. The indexer KL loss (``dsa_indexer_loss_coeff``) and the
+    top-k selection itself are NOT counted: like everywhere else in this file
+    only the model's defining GEMMs enter the estimate, not auxiliary-loss or
```
Comment (alokpathy) [inline megatron/training/training.py]: wrt model's "defining GEMMs", this doesn't seem to be true. The FLOP calculation does include norms (eg line 1020). This seems to be somewhat consistent through the file since MLA FLOP calculations also includes post-norms.

Comment (niyunsheng) [reply megatron/training/training.py]: Good catch. Clarified the docstring in e0dd9463f. No changes to the FLOPs formulas.

File: megatron/training/training.py [hunk 3]
```diff
@@ -1355,52 +1635,113 @@ def _split_spec_part(part):
...
-            gdn_qk_head_dim=args.linear_key_head_dim or 128,
-            gdn_v_head_dim=args.linear_value_head_dim or 128,
-            gdn_num_qk_heads=args.linear_num_key_heads or 16,
-            gdn_num_v_heads=args.linear_num_value_heads or 32,
-            gdn_conv_kernel_dim=args.linear_conv_kernel_dim or 4,
-            gdn_use_gdn2=(args.experimental_attention_variant == "gdn2"),
-            vocab_size=args.padded_vocab_size,
-            mtp_num_layers=mtp_num_layers,
+        return (
+            total_real_tokens_in_batch * dsa_token_linear_term
```
Comment (jaredcasper) [inline megatron/training/training.py:1707]: I think this would be cleaner to separate this out as a "dsa_flops" value or something, so it's clear that you're adding dsa flops to hybrid flops here. (Or if it gets moved a seperate function can be `dsa_flops() + hybrid_flops()`)

File: megatron/training/training.py (hunk 1 above)
Comment (jaredcasper) [reply megatron/training/training.py:824]: How about these new functions go into `megatron/training/flops/dsa.py`, then a follow up can move the rest into `megatron/training/flops/whatever.py`?

File: megatron/training/training.py [hunk 4]
```diff
@@ -1355,52 +1635,113 @@ def _split_spec_part(part):
...
         )
+        layer_counts = get_hybrid_layer_counts(args.hybrid_layer_pattern)
         num_mamba_layers, num_gdn_layers, num_attn_layers, num_mlp_layers, num_moe_layers = (
             itemgetter(Symbols.MAMBA, Symbols.GDN, Symbols.ATTENTION, Symbols.MLP, Symbols.MOE)(
-                get_hybrid_layer_counts(args.hybrid_layer_pattern)
+                layer_counts
             )
         )
 
+        # DSA ('D') layers: absorbed-MLA projections + top-k sparse core
```
Comment (jaredcasper) [inline megatron/training/training.py:1647]: Maybe move this block into a `dsa_flops` function to mirror `hybrid_flops`

File: megatron/training/training.py (hunk 1 above)
Comment (Phlip79) [reply megatron/training/training.py:824]: Can we use the newly created `megatron/training/logging/`? This directory currently only contains `packed_sequence_stats.py`. Alternatively, we can rename logging to stats or flops.

**Review by Phlip79 — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]
